pysoa/server/action/switched.py

```diff
@@ -0,0 +1,122 @@
+
+
+def _len(item):
+    # Safe length that won't raise an error on values that don't support length
+    return getattr(item, '__len__', lambda *_: -1)()
+
+
+@six.add_metaclass(abc.ABCMeta)
+class SwitchedActionMetaClass(type):
+    def __new__(mcs, name, bases, body):
+        """
+        Validate the switch_to_action_map when the class is created, instead of doing it every time the class
+        is instantiated. This identifies problems earlier (on import) and improves performance by not performing this
+        validation every time the action is called.
+        """
+        cls = super(SwitchedActionMetaClass, mcs).__new__(mcs, name, bases, body)
+
+        # noinspection PyUnresolvedReferences
+        if bases[0] is not object and (
+            not cls.switch_to_action_map or
+            not hasattr(cls.switch_to_action_map, '__iter__') or
+            _len(cls.switch_to_action_map) < 2 or
+            any(
+                True for i in cls.switch_to_action_map
+                if not hasattr(i, '__getitem__') or _len(i) != 2 or not is_switch(i[0]) or not callable(i[1])
+            )
+        ):
+            raise ValueError(
+                'Class attribute switch_to_action_map must be an iterable of at least two indexable items, each '
+                'with exactly two indexes, where the first element is a switch and the second element is an action '
+                '(callable).'
+            )
+
+        return cls
+
+
+@six.add_metaclass(SwitchedActionMetaClass)
+class SwitchedAction(object):
+    """
+    A specialized action that defers to other, concrete actions based on request switches. Subclasses must not
+    override any methods and must override `switch_to_action_map`. `switch_to_action_map` should be some iterable
+    object that provides `__len__` (such as a tuple [recommended] or list). Its items must be indexable objects that
+    provide `__len__` (such as a tuple [recommended] or list) and have exactly two elements.
+
+    For each item in `switch_to_action_map`, the first element must be a switch that provides `__int__` (such as an
+    actual integer) or a switch that provides an attribute `value` which, itself, provides `__int__` (or is an int).
+    The second element must be an action, such as an action class (e.g. one that extends `Action`) or any callable
+    that accepts a server settings object and returns a new callable that, itself, accepts an `ActionRequest` object
+    and returns an `ActionResponse` object or raises an `ActionError`.
+
+    `switch_to_action_map` must have at least two items in it. `SwitchedAction` will iterate over that list, checking
+    the first element (switch) to see if it is enabled in the request. If it is, the second item (the action) will be
+    deferred to. If it finds no items whose switches are enabled, it will use the very last action in
+    `switch_to_action_map`. As such, you can treat the last item as a default, and its switch could simply be `0`.
```

Comment:
I would rather we had an explicit default rather than it being the last one - say that "value X means default", and have it be last be a convention rather than required. I might suggest `None` rather than `0` as well.

Reply:
Updated, as discussed here and on Slack.